src/v/cluster/scheduling/leader_balancer_constraints.cc

```diff
@@ -0,0 +1,385 @@
+            }
+        }
+    }
+}
+
+/**
+ *  Used to calculate the initial values for the error this constraint
+ *  is trying to minimize. The goal here is to calculate a per topic
+ *  error(or skew) where the error is zero if the leaders of a topic's
+ *  partitions are evenly distributed on every node. And where the error
+ *  grows to +infinity the more skewed the leadership assignment is to a
+ *  subset of nodes. The equations used can be summarized as;
+ *
+ *  skew[topic_i] = SUM(leaders[node_i, topic_i] - opt[topic_i])^2
+ *  opt[topic_i]  = total_partitions[topic_i] / total_nodes[topic_i]
+ *  where total_nodes is the number of nodes a topic has replicas on.
+ *        total_partitions is the number of partitions the topic has.
+ */
+void even_topic_distributon_constraint::calc_topic_skew() {
+    _topic_skew.clear();
+    _topic_opt_leaders.clear();
+
+    for (const auto& topic : _topic_node_index) {
+        auto topic_partitions = static_cast<double>(
+          _topic_partition_index.at(topic.first));
+        auto topic_replicas = static_cast<double>(
+          _topic_replica_index.at(topic.first).size());
+        auto opt_leaders = topic_partitions
+                           / std::min(topic_replicas, topic_partitions);
+        _topic_opt_leaders[topic.first] = opt_leaders;
+
+        auto& skew = _topic_skew[topic.first];
+        skew = 0;
+
+        for (const auto& node : topic.second) {
+            auto leaders = static_cast<double>(node.second.size());
+
+            skew += pow(leaders - opt_leaders, 2);
+        }
+    }
+}
+
+/*
+ * Compute new error for the given reassignment.
+ */
+double even_topic_distributon_constraint::adjusted_error(
+  double current_error,
+  const topic_id_t& topic_id,
+  const model::broker_shard& from,
+  const model::broker_shard& to) const {
+    // Moving leadership between shards on the same node doesn't change
+    // error for this constraint.
+    if (from.node_id == to.node_id) {
+        return current_error;
+    }
+
+    auto opt_leaders = _topic_opt_leaders.at(topic_id);
+
+    const auto& topic_leaders = _topic_node_index.at(topic_id);
+
+    double from_node_leaders = 0;
+    const auto from_it = topic_leaders.find(from.node_id);
+    if (from_it != topic_leaders.cend()) {
+        from_node_leaders = static_cast<double>(from_it->second.size());
+        current_error -= pow(from_node_leaders - opt_leaders, 2);
+    } else {
+        // If there are no leaders for the topic on the from node
+        // then there is nothing to move and no change to the error.
+        return current_error;
+    }
+
+    double to_node_leaders = 0;
+    const auto to_it = topic_leaders.find(to.node_id);
+    if (to_it != topic_leaders.cend()) {
+        to_node_leaders = static_cast<double>(to_it->second.size());
+        current_error -= pow(to_node_leaders - opt_leaders, 2);
+    }
+    // else there are no leaders assigned to this node.
```

Comment:
```
else return current_error;
```
? 
Like with 'from_it' above?

Reply:
Good catch, this actually needed to be pulled out of the conditional. The math should be correct now.